streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java

```diff
@@ -581,12 +550,13 @@ public GroupAssignment assign(final Cluster metadata, final GroupSubscription gr
 
         final Map<UUID, ClientState> states = new HashMap<>();
         for (final Map.Entry<UUID, ClientMetadata> entry : clientMetadataMap.entrySet()) {
+            final UUID uuid = entry.getKey();
             final ClientState state = entry.getValue().state;
-            states.put(entry.getKey(), state);
+            states.put(uuid, state);
 
-            // Either the active tasks (eager) OR the owned partitions (cooperative) were encoded in the subscription
-            // according to the rebalancing protocol, so convert any partitions in a client to tasks where necessary
-            if (!state.ownedPartitions().isEmpty()) {
+            // this is an optimization: we can't decode the future subscription info's prev tasks, but we can figure
+            // them out from the encoded ownedPartitions
+            if (uuid == futureId && !state.ownedPartitions().isEmpty()) {
```

Comment:
Just to be clear, does this mean we're certain that for non-future members (current or older-versioned ones), the encoded "prevTasks" actually contains all the previous tasks?

I gather this is true from the SubscriptionInfo protcol:
```
    {
      "name": "prevTasks",
      "versions": "1-6",
      "type": "[]TaskId"
    }
```

But then, I'm a little mystified by the prior comment... why would "active tasks" not have been encoded with the cooperative protocol?

Reply:
Let me give a quick history lesson to clarify:
By adding the `ownedPartitions` field to the subscription we were able to remove the encoded prevTasks and avoid duplicating info we could get from the new field for COOPERATIVE members. Note the `ownedPartitions` are the source of truth and may differ from the prevTasks that would have been encoded in edge cases (eg topic deletion, partitions lost)
In hindsight, aka a few weeks back when I started looking at the assignor code again, I realized this was just unnecessary and likely to lead to more trouble than it solves. So, now we just encode all tasks in the offset map regardless of rebalance protocol.
Ok, while explaining that I now realize we obviously still need to fill in the `prevTasks` from the `ownedPartitions` for members on 2.5/2.6 -- thanks Socrates :P